File: examples/c/mcc172/multi_hat_synchronous_scan/multi_hat_synchronous_scan.c
/*****************************************************************************

    MCC 172 Functions Demonstrated:
        mcc172_trigger_config
        mcc172_a_in_clock_config_write
        mcc172_a_in_clock_config_read
        mcc172_a_in_scan_start
        mcc172_a_in_scan_status
        mcc172_a_in_scan_read
        mcc172_a_in_scan_stop

    Purpose:
        Get synchronous data from multiple MCC 172 devices.

    Description:
        This example demonstrates acquiring data synchronously from multiple
        MCC 172 devices.  This is done using the shared clock and trigger
        options.  An external trigger source must be provided to the TRIG
        terminal on the master MCC 172 device.  The clock and trigger on the
        master device are configured for SOURCE_MASTER and the remaining devices
        are configured for SOURCE_SLAVE.

*****************************************************************************/
#include "../../daqhats_utils.h"
#include <math.h>

#define DEVICE_COUNT 2
#define MASTER 0

#define CURSOR_SAVE "\x1b[s"
#define CURSOR_RESTORE "\x1b[u"

int get_hat_addresses(uint8_t address[]);

double calc_rms(double* data, uint8_t channel, uint8_t num_channels,
    uint32_t num_samples_per_channel)
{
    double value;
    uint32_t i;
    uint32_t index;
    
    value = 0.0;
    for (i = 0; i < num_samples_per_channel; i++)
    {
        index = (i * num_channels) + channel;
        value += (data[index] * data[index]) / num_samples_per_channel;
    }
    
    return sqrt(value);
}


int main(void)
{
    // HAT device addresses - determined at runtime
    uint8_t address[DEVICE_COUNT];

    // mcc172_a_in_scan_start() variables
    uint8_t chan_mask[DEVICE_COUNT] = {
        CHAN0 | CHAN1,
        CHAN0 | CHAN1
    };
    uint32_t options = OPTS_EXTTRIGGER;
    
    uint32_t samples_per_channel = 10240;
    double sample_rate = 10240; // Samples per second

    // mcc172_trigger_config() variable
    uint8_t trigger_mode = TRIG_RISING_EDGE;

    int mcc172_num_channels = mcc172_info()->NUM_AI_CHANNELS;

    // mcc172_a_in_scan_status() and mcc172_a_in_scan_read() variables
    uint16_t scan_status[DEVICE_COUNT] = {0};
    int buffer_size = samples_per_channel * mcc172_num_channels;
    double data[DEVICE_COUNT][buffer_size];
    uint32_t samples_read[DEVICE_COUNT] = {0};
    int32_t samples_to_read = 1000;
    double timeout = 5;  //seconds
    uint32_t samples_available = 0;

    int result = 0;
    int chan_count[DEVICE_COUNT];
    int chans[DEVICE_COUNT][mcc172_num_channels];
    int device = 0;
    int i = 0;
    int total_samples_read[DEVICE_COUNT] = {0};
    const int data_display_line_count = DEVICE_COUNT * 4;
    uint16_t is_running = 0;
    uint16_t is_triggered = 0;
    uint16_t scan_status_all = 0;

    char chan_display[DEVICE_COUNT][32];
    char options_str[256];
    char display_string[256] = "";
    char data_display_output[1024] = "";
    char c;
    uint8_t synced;
    uint8_t clock_source;

    double actual_sample_rate = 0.0;

    // Determine the addresses of the devices to be used
    if (get_hat_addresses(address) != 0)
    {
        return -1;
    }

    convert_options_to_string(options, options_str);
    
    for (device = 0; device < DEVICE_COUNT; device++)
    {
        // Open a connection to each device
        result = mcc172_open(address[device]);
        STOP_ON_ERROR(result);

        // channel_mask is used by the library function mcc172_a_in_scan_start.
        // The functions below parse the mask for display purposes.
        chan_count[device] = convert_chan_mask_to_array(chan_mask[device], 
            chans[device]);
        convert_chan_mask_to_string(chan_mask[device], chan_display[device]);
        
        // Configure the trigger
        result = mcc172_trigger_config(address[device], 
            (device == MASTER) ? SOURCE_MASTER : SOURCE_SLAVE,
            trigger_mode);
        STOP_ON_ERROR(result);
        
        // Configure the clock (slaves only)
        if (device != MASTER)
        {
            result = mcc172_a_in_clock_config_write(address[device], 
                SOURCE_SLAVE, sample_rate);
            STOP_ON_ERROR(result);
        }
    }

    // Configure the master clock last so the clocks are synchronized
    result = mcc172_a_in_clock_config_write(address[MASTER], 
        SOURCE_MASTER, sample_rate);
    STOP_ON_ERROR(result);

    // Wait for sync to complete
    do
    {
        result = mcc172_a_in_clock_config_read(address[MASTER], &clock_source,
            &actual_sample_rate, &synced);
        STOP_ON_ERROR(result);
        usleep(5000);
    } while (synced == 0);
    

    printf("\nMCC 172 multiple device example using shared clock and "
        "trigger options\n");
    printf("    Functions demonstrated:\n");
    printf("      mcc172_trigger_config\n");
    printf("      mcc172_a_in_clock_config_write\n");
    printf("      mcc172_a_in_clock_config_read\n");
    printf("      mcc172_a_in_scan_start\n");
    printf("      mcc172_a_in_scan_status\n");
    printf("      mcc172_a_in_scan_read\n");
    printf("      mcc172_a_in_scan_stop\n");
    printf("    Samples per channel: %d\n", samples_per_channel);
    printf("    Requested Sample Rate: %.3f Hz\n", sample_rate);
    printf("    Actual Sample Rate: %.3f Hz\n", actual_sample_rate);
    convert_trigger_mode_to_string(trigger_mode, display_string);
    printf("    Trigger mode: %s\n", display_string);

    for (device = 0; device < DEVICE_COUNT; device++)
    {
        printf("    MCC 172 %d:\n", device);
        printf("      Address: %d\n", address[device]);
        printf("      Channels: %s\n", chan_display[device]);
    }

    printf("\nConnect a trigger source to the TRIG input terminal on "
        "device at address %d.\n", address[0]);

    printf("\nPress 'Enter' to continue\n");

    scanf("%c", &c);

    // Start the scans
    for (device = 0; device < DEVICE_COUNT; device++)
    {
        result = mcc172_a_in_scan_start(address[device], chan_mask[device],
            samples_per_channel, options);
        STOP_ON_ERROR(result);
    }

    printf("Waiting for trigger ... Press 'Enter' to abort\n\n");

    do
    {
        usleep(10000);
        // Check the master status in a loop to determine when the trigger occurs
        result = mcc172_a_in_scan_status(address[MASTER], &scan_status[MASTER],
                                         &samples_available);
        STOP_ON_ERROR(result);
        
        is_running = (scan_status[MASTER] & STATUS_RUNNING);
        is_triggered = (scan_status[MASTER] & STATUS_TRIGGERED);
        
    }
    while (is_running && !is_triggered && !enter_press());

    if (is_running && is_triggered)
    {
        printf("Acquiring data ... Press 'Enter' to abort\n\n");
        // Create blank lines where the data will be displayed.
        for (i = 0; i <= data_display_line_count; i++)
        {
            printf("\n");
        }
        // Move the cursor up to the start of the data display
        printf("\x1b[%dA", data_display_line_count + 1);
        // Save the cursor position
        printf(CURSOR_SAVE);
    }
    else
    {
        printf("Aborted\n\n");
        is_running = 0;
    }

    while (is_running)
    {
        for (device = 0; device < DEVICE_COUNT; device++)
        {
            // Read data
            result = mcc172_a_in_scan_read(address[device], 
                &scan_status[device], samples_to_read, timeout, data[device],
                buffer_size, &samples_read[device]);
            STOP_ON_ERROR(result);
            // Check for overrun on any one device
            scan_status_all |= scan_status[device];
            // Verify the status of all devices is running
            is_running &= (scan_status[device] & STATUS_RUNNING);
        }

        if ((scan_status_all & STATUS_HW_OVERRUN) == STATUS_HW_OVERRUN)
        {
            fprintf(stderr, "\nError: Hardware overrun\n");
            break;
        }
        if ((scan_status_all & STATUS_BUFFER_OVERRUN) == STATUS_BUFFER_OVERRUN)
        {
            fprintf(stderr, "\nError: Buffer overrun\n");
            break;
        }

        // Restore the cursor position to the start of the data display
        printf(CURSOR_RESTORE);
        
        for (device = 0; device < DEVICE_COUNT; device++)
        {
            strcpy(data_display_output, "");

            sprintf(display_string, "HAT %d:\n", device);
            strcat(data_display_output, display_string);

            // Display the header row for the data table
            strcat(data_display_output, "  Samples Read    Scan Count");
            for (i = 0; i < chan_count[device]; i++)
            {
                sprintf(display_string, "     Channel %d", chans[device][i]);
                strcat(data_display_output, display_string);
            }
            strcat(data_display_output, "\n");

            // Display the sample count information for the device
            total_samples_read[device] += samples_read[device];
            sprintf(display_string, "%14d%14d", samples_read[device],
                total_samples_read[device]);
            strcat(data_display_output, display_string);

            // Display the data for all active channels
            if (samples_read[device] > 0)
            {
                for (i = 0; i < chan_count[device]; i++)
                {
                    // Calculate and display RMS voltage of the input data
                    sprintf(display_string, "%9.3f Vrms",
                        calc_rms(data[device], i, chan_count[device],
                            samples_read[device]));
                    strcat(data_display_output, display_string);
                }
            }

            strcat(data_display_output, "\n\n");
            printf(data_display_output);
        }

        fflush(stdout);

        if (enter_press())
        {
            printf("Aborted\n\n");
            break;
        }
    }

stop:
    // Stop and cleanup
    for (device = 0; device < DEVICE_COUNT; device++)
    {
        result = mcc172_a_in_scan_stop(address[device]);
        print_error(result);
        result = mcc172_a_in_scan_cleanup(address[device]);
        print_error(result);
        
        // Restore clock and trigger to local source
        result = mcc172_a_in_clock_config_write(address[device], 
            SOURCE_LOCAL, sample_rate);
        print_error(result);
        
        result = mcc172_trigger_config(address[device], 
            SOURCE_LOCAL, trigger_mode);
        print_error(result);

        result = mcc172_close(address[device]);
        print_error(result);
    }

    return 0;
}

/* This function gets the addresses of the MCC 172 devices to be used. */
int get_hat_addresses(uint8_t address[])
{
    int return_val = -1;
    int hat_count = 0;
    int device = 0;
    int i = 0;
    int address_int;

    int valid = 0;

    hat_count = hat_list(HAT_ID_MCC_172 , NULL);

    if (hat_count >= DEVICE_COUNT)
    {
        struct HatInfo* hats = (struct HatInfo*)malloc(
            hat_count * sizeof(struct HatInfo));
        hat_list(HAT_ID_MCC_172 , hats);

        if (hat_count == DEVICE_COUNT)
        {
            for (device = 0; device < DEVICE_COUNT; device++)
            {
                address[device] = hats[device].address;
            }
        }
        else
        {
            for (i = 0; i < hat_count; i++)
            {
                printf("Address %d: %s\n", hats[i].address, 
                    hats[i].product_name);
            }

            for (device = 0; device < DEVICE_COUNT; device++)
            {
                valid = 0;
                while (!valid)
                {
                    printf("\n Enter address for HAT device %d: ", device);
                    scanf("%d", &address_int);

                    // Check if the address exists in the hats array
                    for (i = 0; i < hat_count; i++)
                    {
                        if (hats[i].address == address_int)
                            valid = 1;
                    }
                    // Make sure the address was not previously selected
                    for (i = 0; i < device; i++)
                    {
                        if (address[i] == address_int)
                            valid = 0;
                    }

                    if (valid)
                    {
                        address[device] = (uint8_t)address_int;
                    }
                    else
                    {
                        printf("Invalid address - try again");
                    }
                }
            }
            flush_stdin();
        }
        free(hats);
        return_val = 0;
    }
    else
    {
        fprintf(stderr, "Error: This example requires %d MCC 172 devices - "
            "found %d\n", DEVICE_COUNT, hat_count);
    }

    return return_val;
}

File: examples/python/mcc172/multi_hat_synchronous_scan.py
#!/usr/bin/env python
#  -*- coding: utf-8 -*-
"""
    MCC 172 Functions Demonstrated:
        mcc172.trigger_config
        mcc172.iepe_config_write
        mcc172.a_in_clock_config_write
        mcc172.a_in_clock_config_read
        mcc172.a_in_scan_start
        mcc172.a_in_scan_read
        mcc172.a_in_scan_stop

    Purpose:
    Get synchronous data from multiple MCC 172 HAT devices.

    Description:
        This example demonstrates acquiring data synchronously from multiple
        MCC 172 HAT devices.  This is done using the shared clock and
        trigger options.  An external trigger source must be provided to the
        TRIG terminal on the master MCC 172 HAT device.  The EXTTRIGGER scan
        option is set on all devices.
"""
from __future__ import print_function
from sys import stdout, version_info
from math import sqrt
from time import sleep
from daqhats import (hat_list, mcc172, OptionFlags, HatIDs, TriggerModes,
                     HatError, SourceType)
from daqhats_utils import (enum_mask_to_string, chan_list_to_mask,
                           validate_channels)

# Constants
DEVICE_COUNT = 2
MASTER = 0
CURSOR_SAVE = "\x1b[s"
CURSOR_RESTORE = "\x1b[u"
CURSOR_BACK_2 = '\x1b[2D'
ERASE_TO_END_OF_LINE = '\x1b[0K'

def get_iepe():
    """
    Get IEPE enable from the user.
    """

    while True:
        # Wait for the user to enter a response
        message = "IEPE enable [y or n]?  "
        if version_info.major > 2:
            response = input(message)
        else:
            response = raw_input(message)

        # Check for valid response
        if (response == "y") or (response == "Y"):
            return 1
        elif (response == "n") or (response == "N"):
            return 0
        else:
            # Ask again.
            print("Invalid response.")

def main(): # pylint: disable=too-many-locals, too-many-statements, too-many-branches
    """
    This function is executed automatically when the module is run directly.
    """
    hats = []
    # Define the channel list for each HAT device
    chans = [
        {0, 1},
        {0, 1}
    ]
    # Define the options for each HAT device
    options = [
        OptionFlags.EXTTRIGGER,
        OptionFlags.EXTTRIGGER
    ]
    samples_per_channel = 10000
    sample_rate = 10240.0  # Samples per second
    trigger_mode = TriggerModes.RISING_EDGE

    try:
        # Get an instance of the selected hat device object.
        hats = select_hat_devices(HatIDs.MCC_172, DEVICE_COUNT)

        # Validate the selected channels.
        for i, hat in enumerate(hats):
            validate_channels(chans[i], hat.info().NUM_AI_CHANNELS)

        # Turn on IEPE supply?
        iepe_enable = get_iepe()

        for i, hat in enumerate(hats):
            for channel in chans[i]:
                # Configure IEPE.
                hat.iepe_config_write(channel, iepe_enable)
            if hat.address() != MASTER:
                # Configure the slave clocks.
                hat.a_in_clock_config_write(SourceType.SLAVE, sample_rate)
                # Configure the trigger.
                hat.trigger_config(SourceType.SLAVE, trigger_mode)

        # Configure the master clock and start the sync.
        hats[MASTER].a_in_clock_config_write(SourceType.MASTER, sample_rate)
        synced = False
        while not synced:
            (_source_type, actual_rate, synced) = \
                hats[MASTER].a_in_clock_config_read()
            if not synced:
                sleep(0.005)

        # Configure the master trigger.
        hats[MASTER].trigger_config(SourceType.MASTER, trigger_mode)

        print('MCC 172 multiple HAT example using external clock and',
              'external trigger options')
        print('    Functions demonstrated:')
        print('         mcc172.trigger_mode')
        print('         mcc172.a_in_clock_config_write')
        print('         mcc172.a_in_clock_config_read')
        print('         mcc172.a_in_scan_start')
        print('         mcc172.a_in_scan_read')
        print('         mcc172.a_in_scan_stop')
        print('         mcc172.a_in_scan_cleanup')
        print('    IEPE power: ', end='')
        if iepe_enable == 1:
            print('on')
        else:
            print('off')
        print('    Samples per channel:', samples_per_channel)
        print('    Requested Sample Rate: {:.3f} Hz'.format(sample_rate))
        print('    Actual Sample Rate: {:.3f} Hz'.format(actual_rate))
        print('    Trigger type:', trigger_mode.name)

        for i, hat in enumerate(hats):
            print('    HAT {}:'.format(i))
            print('      Address:', hat.address())
            print('      Channels: ', end='')
            print(', '.join([str(chan) for chan in chans[i]]))
            options_str = enum_mask_to_string(OptionFlags, options[i])
            print('      Options:', options_str)

        print('\n*NOTE: Connect a trigger source to the TRIG input terminal on HAT 0.')

        try:
            input("\nPress 'Enter' to continue")
        except (NameError, SyntaxError):
            pass

        # Start the scan.
        for i, hat in enumerate(hats):
            chan_mask = chan_list_to_mask(chans[i])
            hat.a_in_scan_start(chan_mask, samples_per_channel, options[i])

        print('\nWaiting for trigger ... Press Ctrl-C to stop scan\n')

        try:
            # Monitor the trigger status on the master device.
            wait_for_trigger(hats[MASTER])
            # Read and display data for all devices until scan completes
            # or overrun is detected.
            read_and_display_data(hats, chans)

        except KeyboardInterrupt:
            # Clear the '^C' from the display.
            print(CURSOR_BACK_2, ERASE_TO_END_OF_LINE, '\nAborted\n')

    except (HatError, ValueError) as error:
        print('\n', error)

    finally:
        for hat in hats:
            hat.a_in_scan_stop()
            hat.a_in_scan_cleanup()


def wait_for_trigger(hat):
    """
    Monitor the status of the specified HAT device in a loop until the
    triggered status is True or the running status is False.

    Args:
        hat (mcc172): The mcc172 HAT device object on which the status will
            be monitored.

    Returns:
        None

    """
    # Read the status only to determine when the trigger occurs.
    is_running = True
    is_triggered = False
    while is_running and not is_triggered:
        status = hat.a_in_scan_status()
        is_running = status.running
        is_triggered = status.triggered

def calc_rms(data, channel, num_channels, num_samples_per_channel):
    """ Calculate RMS value from a block of samples. """
    value = 0.0
    index = channel
    for _i in range(num_samples_per_channel):
        value += (data[index] * data[index]) / num_samples_per_channel
        index += num_channels

    return sqrt(value)

def read_and_display_data(hats, chans):
    """
    Reads data from the specified channels on the specified DAQ HAT devices
    and updates the data on the terminal display.  The reads are executed in a
    loop that continues until either the scan completes or an overrun error
    is detected.

    Args:
        hats (list[mcc172]): A list of mcc172 HAT device objects.
        chans (list[int][int]): A 2D list to specify the channel list for each
            mcc172 HAT device.

    Returns:
        None

    """
    samples_to_read = 1000
    timeout = 5  # Seconds
    samples_per_chan_read = [0] * DEVICE_COUNT
    total_samples_per_chan = [0] * DEVICE_COUNT
    is_running = True

    # Since the read_request_size is set to a specific value, a_in_scan_read()
    # will block until that many samples are available or the timeout is
    # exceeded.

    # Create blank lines where the data will be displayed
    for _ in range(DEVICE_COUNT * 4 + 1):
        print('')
    # Move the cursor up to the start of the data display.
    print('\x1b[{0}A'.format(DEVICE_COUNT * 4 + 1), end='')
    print(CURSOR_SAVE, end='')

    while True:
        data = [None] * DEVICE_COUNT
        # Read the data from each HAT device.
        for i, hat in enumerate(hats):
            read_result = hat.a_in_scan_read(samples_to_read, timeout)
            data[i] = read_result.data
            is_running &= read_result.running
            samples_per_chan_read[i] = int(len(data[i]) / len(chans[i]))
            total_samples_per_chan[i] += samples_per_chan_read[i]

            if read_result.buffer_overrun:
                print('\nError: Buffer overrun')
                break
            if read_result.hardware_overrun:
                print('\nError: Hardware overrun')
                break

        print(CURSOR_RESTORE, end='')

        # Display the data for each HAT device
        for i, hat in enumerate(hats):
            print('HAT {0}:'.format(i))

            # Print the header row for the data table.
            print('  Samples Read    Scan Count', end='')
            for chan in chans[i]:
                print('     Channel', chan, end='')
            print('')

            # Display the sample count information.
            print('{0:>14}{1:>14}'.format(samples_per_chan_read[i],
                                          total_samples_per_chan[i]), end='')

            # Display the data for all selected channels
            #for chan_idx in range(len(chans[i])):
            #    sample_idx = ((samples_per_chan_read[i] * len(chans[i]))
            #                  - len(chans[i]) + chan_idx)
            #    print('{:>12.5f} V'.format(data[i][sample_idx]), end='')

            # Display the RMS voltage for each channel.
            if samples_per_chan_read[i] > 0:
                for channel in chans[i]:
                    value = calc_rms(data[i], channel, len(chans[i]),
                                     samples_per_chan_read[i])
                    print('{:10.5f}'.format(value), 'Vrms ',
                          end='')
                stdout.flush()
            print('\n')

        #stdout.flush()

        if not is_running:
            break


def select_hat_devices(filter_by_id, number_of_devices):
    """
    This function performs a query of available DAQ HAT devices and determines
    the addresses of the DAQ HAT devices to be used in the example.  If the
    number of HAT devices present matches the requested number of devices,
    a list of all mcc172 objects is returned in order of address, otherwise the
    user is prompted to select addresses from a list of displayed devices.

    Args:
        filter_by_id (int): If this is :py:const:`HatIDs.ANY` return all DAQ
            HATs found.  Otherwise, return only DAQ HATs with ID matching this
            value.
        number_of_devices (int): The number of devices to be selected.

    Returns:
        list[mcc172]: A list of mcc172 objects for the selected devices
        (Note: The object at index 0 will be used as the master).

    Raises:
        HatError: Not enough HAT devices are present.

    """
    selected_hats = []

    # Get descriptors for all of the available HAT devices.
    hats = hat_list(filter_by_id=filter_by_id)
    number_of_hats = len(hats)

    # Verify at least one HAT device is detected.
    if number_of_hats < number_of_devices:
        error_string = ('Error: This example requires {0} MCC 172 HATs - '
                        'found {1}'.format(number_of_devices, number_of_hats))
        raise HatError(0, error_string)
    elif number_of_hats == number_of_devices:
        for i in range(number_of_devices):
            selected_hats.append(mcc172(hats[i].address))
    else:
        # Display available HAT devices for selection.
        for hat in hats:
            print('Address ', hat.address, ': ', hat.product_name, sep='')
        print('')

        for device in range(number_of_devices):
            valid = False
            while not valid:
                input_str = 'Enter address for HAT device {}: '.format(device)
                address = int(input(input_str))

                # Verify the selected address exists.
                if any(hat.address == address for hat in hats):
                    valid = True
                else:
                    print('Invalid address - try again')

                # Verify the address was not previously selected
                if any(hat.address() == address for hat in selected_hats):
                    print('Address already selected - try again')
                    valid = False

                if valid:
                    selected_hats.append(mcc172(address))

    return selected_hats


if __name__ == '__main__':
    # This will only be run when the module is called directly.
    main()
